### PR TITLE
[CALCITE-6236] EnumerableBatchNestedLoopJoin::estimateRowCount returns wrong value

### PR DESCRIPTION
**Pull request by kramerul**
This is another approach to fix [CALCITE-6236] using the rowCount of the original relation.

### CODE REVIEW

**Comment by kramerul** [comment]
We are still interested in getting this PR merged.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableBatchNestedLoopJoinTest.java [hunk 1]
```diff
@@ -226,6 +226,7 @@ class EnumerableBatchNestedLoopJoinTest {
             + "join locations l on e.empid <> l.empid and d.deptno = l.empid")
         .withHook(Hook.PLANNER, (Consumer<RelOptPlanner>) planner -> {
           planner.removeRule(EnumerableRules.ENUMERABLE_CORRELATE_RULE);
+          planner.removeRule(EnumerableRules.ENUMERABLE_JOIN_RULE);
```
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/enumerable/EnumerableBatchNestedLoopJoinTest.java:229]: is this related to the problem addressed?

**Review by mihaibudiu — CHANGES_REQUESTED** [review]
There are no tests...
There was a healthy discussion in the JIRA, hopefully some of the people who commented can also review.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableBatchNestedLoopJoinTest.java (hunk 1 above)
Comment (kramerul) [reply core/src/test/java/org/apache/calcite/test/enumerable/EnumerableBatchNestedLoopJoinTest.java:229]: It's related to the test. When the rule is not removed, the planner creates a `EnumerableHashJoin` instead of a `EnumerableBatchNestedLoopJoin` because the cost function is affected by this change.

**Comment by rubenada** [comment]
Sorry but the proposed approach (storing "originalRowCount" on the BNLJ) seems to go against some of the concepts discussed [on the Jira](https://issues.apache.org/jira/browse/CALCITE-6236?focusedCommentId=17813606&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17813606):
```
pre-storing selectivity or other kind of metadata information in the operator also feels like we are going to the wrong direction
```

This is certainly a tricky one. IMO if we can't reach consensus about this topic, it'd not be a critical problem since each downstream project can always define their own `MyRelMdRowCount` handler and put in there whatever logic the require to refine BNLJ rowCount estimation.

**Comment by kramerul** [comment]
I will close this PR. We will use a fix inside our application.
